File: python/selene_isru/modules/thermal.py
from __future__ import annotations

import math
from typing import Any

from ..constants import c


def sec_sub_j_per_kg(chi_ice: float, cp_reg_cold: float, tpsr: float, tsub: float) -> float:
    return (1 / chi_ice) * cp_reg_cold * (tsub - tpsr) + c("dHsub_ice")


def simulate_thermal(params: dict[str, Any]) -> dict[str, float | None]:
    temp = params["Tsub"] if params["site"] == "polar" else 300
    conductivity_w_per_mk = params["kc"] + params["kr"] * temp**3
    knudsen_d_m2_per_s = (2 / 3) * params["rPore"] * math.sqrt((8 * c("R") * temp) / (math.pi * c("M_H2O")))

    return {
        "secSub_JPerKg": (
            sec_sub_j_per_kg(params["chiIce"], params["cpRegCold"], params["Tpsr"], params["Tsub"])
            if params["site"] == "polar"
            else None
        ),
        "knudsenD_M2PerS": knudsen_d_m2_per_s,
        "conductivity_WPerMK": conductivity_w_per_mk,
    }

File: packages/engine/src/modules/thermal.ts
import { PHYSICAL_CONSTANTS } from "../constants";
import type { SimParams, SiteMode } from "../types";

export interface ThermalOutput {
  secSub_JPerKg: number | null;
  knudsenD_M2PerS: number;
  conductivity_WPerMK: number;
}

export function secSubJPerKg(chiIce: number, cpRegCold: number, Tpsr: number, Tsub: number): number {
  return (1 / chiIce) * cpRegCold * (Tsub - Tpsr) + PHYSICAL_CONSTANTS.dHsub_ice.value;
}

export function simulateThermal(params: SimParams): ThermalOutput {
  const T = params.site === "polar" ? params.Tsub : 300;
  const conductivity_WPerMK = params.kc + params.kr * T ** 3;
  const knudsenD_M2PerS =
    (2 / 3) *
    params.rPore *
    Math.sqrt((8 * PHYSICAL_CONSTANTS.R.value * T) / (Math.PI * PHYSICAL_CONSTANTS.M_H2O.value));

  return {
    secSub_JPerKg:
      params.site === "polar"
        ? secSubJPerKg(params.chiIce, params.cpRegCold, params.Tpsr, params.Tsub)
        : null,
    knudsenD_M2PerS,
    conductivity_WPerMK
  };
}

export function thermalSiteTemperature(site: SiteMode, Tsub: number): number {
  return site === "polar" ? Tsub : 300;
}
